src/Evolution/Initialization/GrTagsForHydro.hpp

```diff
@@ -94,9 +100,33 @@ struct GrTagsForHydro {
 
     // Set initial data from analytic solution
     GrVars gr_vars{num_grid_points};
-    gr_vars.assign_subset(evolution::Initialization::initial_data(
-        Parallel::get<::Tags::AnalyticSolutionOrData>(cache), inertial_coords,
-        initial_time, typename GrVars::tags_list{}));
+    if constexpr (db::tag_is_retrievable_v<
+                      evolution::initial_data::Tags::InitialData,
+                      db::DataBox<DbTagsList>>) {
+      using derived_classes =
+          tmpl::at<typename Metavariables::factory_creation::factory_classes,
+                   evolution::initial_data::InitialData>;
+      call_with_dynamic_type<void, derived_classes>(
+          &db::get<evolution::initial_data::Tags::InitialData>(box),
+          [&gr_vars, &initial_time,
+           &inertial_coords](const auto* const data_or_solution) {
+            gr_vars.assign_subset(evolution::Initialization::initial_data(
+                *data_or_solution, inertial_coords, initial_time,
+                typename GrVars::tags_list{}));
+          });
+    } else if constexpr (db::tag_is_retrievable_v<
+                             ::Tags::AnalyticSolutionOrData,
+                             db::DataBox<DbTagsList>>) {
+      gr_vars.assign_subset(evolution::Initialization::initial_data(
+          Parallel::get<::Tags::AnalyticSolutionOrData>(cache), inertial_coords,
+          initial_time, typename GrVars::tags_list{}));
+    } else {
+      ERROR(
+          "Either ::Tags::AnalyticSolutionOrData or "
+          "evolution::initial_data::Tags::InitialData must be in the "
```

Comment:
Can this be `static_assert`ed instead of runtime?  I don't know if this is ever instantiated with "bad" values but never run.